dali/kernels/slice/slice_flip_normalize_permute_pad_cuda_impl.cuh

```diff
@@ -31,122 +32,153 @@
 namespace dali {
 namespace kernels {
 
+
 namespace detail {
 
 template <int Dims>
 struct SampleDesc {
   void *__restrict__ out;
   const void *__restrict__ in;
-  TensorShape<Dims> in_strides;
-  TensorShape<Dims> out_strides;
+
   TensorShape<Dims> out_shape;
-  TensorShape<Dims> padded_out_shape;
-  float padding_val;
+  TensorShape<Dims> in_shape;
+  TensorShape<Dims> anchor;
+
+  const void *__restrict__ fill_values;
+  const float *__restrict__ norm_add;
+  const float *__restrict__ norm_mul;
+  int channel_dim;
+  bool need_pad;
+  bool need_flip;
+
+  fast_div<uint64_t> out_strides[Dims];
+  TensorShape<Dims> in_strides;
 };
 
 struct BlockDesc {
   int sampleIdx;
-  size_t offset;
-  size_t size;
+  uint64_t offset;
+  uint64_t size;
 };
 
-template <typename OutputType, typename InputType, unsigned Dims, bool should_normalize>
-__device__ inline void SliceFlipNormalizePermutePadFunc(OutputType *__restrict__ out,
-                                                        const InputType *__restrict__ in,
-                                                        const int64_t *out_strides,
-                                                        const int64_t *in_strides,
-                                                        const int64_t *out_shape,
-                                                        const int64_t *padded_out_shape,
-                                                        bool should_pad,
-                                                        unsigned norm_dim,
-                                                        const float *norm_add,
-                                                        const float *norm_mul,
-                                                        OutputType padding_val,
-                                                        size_t offset, size_t block_end) {
-  if (Dims > 1 && !should_normalize &&
-      out_strides[Dims - 1] == in_strides[Dims - 1] &&
-      out_shape[Dims - 1] == padded_out_shape[Dims - 1]) {
-    const unsigned NextDims = Dims > 1 ? Dims - 1 : 1;
-    SliceFlipNormalizePermutePadFunc<OutputType, InputType, NextDims, should_normalize>(
-        out, in, out_strides, in_strides, out_shape, padded_out_shape,
-        should_pad, norm_dim, norm_add, norm_mul, padding_val, offset, block_end);
+DALI_HOST_DEV DALI_FORCEINLINE bool is_out_of_bounds(int64_t idx, int64_t data_extent) {
+  // check idx < 0 and idx >= data_extent at once
+  return static_cast<uint64_t>(idx) >= static_cast<uint64_t>(data_extent);
```

Comment:
Why are you casting? Both values are int64_t so no signed-unsigned comparison.

Reply:
the comment explains it. The casting to `uint64_t` is a trick to check 0 <= idx <= data_extent in one comparison

Reply:
So you are assume that even if it is negative after casting it will be still bigger than  data_extent.

Reply:
yes, it will because of the range of int64_t. This trick was actually suggested by @mzient.